sdks/java/extensions/ml/src/main/java/org/apache/beam/sdk/extensions/ml/RecommendationAIPredict.java

```diff
@@ -0,0 +1,155 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.ml;
+
+import com.google.api.client.json.GenericJson;
+import com.google.api.gax.rpc.ApiException;
+import com.google.auto.value.AutoValue;
+import com.google.cloud.recommendationengine.v1beta1.PlacementName;
+import com.google.cloud.recommendationengine.v1beta1.PredictResponse;
+import com.google.cloud.recommendationengine.v1beta1.PredictionServiceClient;
+import com.google.cloud.recommendationengine.v1beta1.UserEvent;
+import com.google.protobuf.util.JsonFormat;
+import java.io.IOException;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.json.JSONObject;
+
+/**
+ * A {@link PTransform} using the Recommendations AI API (https://cloud.google.com/recommendations).
+ * Takes an input {@link PCollection} of {@link GenericJson}s and creates {@link
+ * PredictResponse.PredictionResult}s.
+ *
+ * <p>It is possible to provide a catalog name to which you want to add the user event (defaults to
+ * "default_catalog"). It is possible to provide a event store to which you want to add the user
+ * event (defaults to "default_event_store"). A placement id for the recommendation engine placement
+ * to be used.
+ */
+@AutoValue
+@SuppressWarnings({"nullness"})
+public abstract class RecommendationAIPredict
+    extends PTransform<PCollection<GenericJson>, PCollectionTuple> {
+
+  /** @return ID of Google Cloud project to be used for creating catalog items. */
+  public abstract @Nullable String projectId();
+
+  /** @return Name of the catalog where the catalog items will be created. */
+  public abstract @Nullable String catalogName();
+
+  /** @return Name of the event store where the user events will be created. */
+  public abstract @Nullable String eventStore();
+
+  /** @return ID of the recommendation engine placement. */
+  public abstract String placementId();
+
+  public static final TupleTag<PredictResponse.PredictionResult> SUCCESS_TAG =
+      new TupleTag<PredictResponse.PredictionResult>() {};
+
+  public static final TupleTag<UserEvent> FAILURE_TAG = new TupleTag<UserEvent>() {};
+
+  @AutoValue.Builder
+  public abstract static class Builder {
```

Comment:
you can remove `public` from this builder so it can only be used within the package